v2/fake/fake.go

```diff
@@ -249,60 +249,219 @@ func UnexpectedActionError() error {
 	return errors.New("Unexpected action")
 }
 
-// CatalogReaction is sent as the response to GetCatalog requests.
+// CatalogReactionInterface is sent as the response to GetCatalog requests.
+type CatalogReactionInterface interface {
+	react() (*v2.CatalogResponse, error)
+}
+
 type CatalogReaction struct {
 	Response *v2.CatalogResponse
 	Error    error
 }
 
-// ProvisionReaction is sent as the response ProvisionInstance requests.
+func (r *CatalogReaction) react() (*v2.CatalogResponse, error) {
+	if r == nil {
+		return nil, UnexpectedActionError()
+	}
+	return r.Response, r.Error
+}
+
+type DynamicCatalogReaction func() (*v2.CatalogResponse, error)
+
+func (r DynamicCatalogReaction) react() (*v2.CatalogResponse, error) {
+	return r()
+}
+
+// ProvisionReactionInterface is sent as the response ProvisionInstance requests.
+type ProvisionReactionInterface interface {
+	react(*v2.ProvisionRequest) (*v2.ProvisionResponse, error)
+}
+
 type ProvisionReaction struct {
 	Response *v2.ProvisionResponse
 	Error    error
 }
 
-// UpdateInstanceReaction is sent as the response UpdateInstance requests.
+func (r *ProvisionReaction) react(_ *v2.ProvisionRequest) (*v2.ProvisionResponse, error) {
+	if r == nil {
+		return nil, UnexpectedActionError()
+	}
+	return r.Response, r.Error
+}
+
+type DynamicProvisionReaction func(*v2.ProvisionRequest) (*v2.ProvisionResponse, error)
+
+func (r DynamicProvisionReaction) react(req *v2.ProvisionRequest) (*v2.ProvisionResponse, error) {
+	return r(req)
+}
+
+// UpdateInstanceReactionInterface is sent as the response UpdateInstance requests.
+type UpdateInstanceReactionInterface interface {
+	react(*v2.UpdateInstanceRequest) (*v2.UpdateInstanceResponse, error)
+}
+
 type UpdateInstanceReaction struct {
 	Response *v2.UpdateInstanceResponse
 	Error    error
 }
 
-// DeprovisionReaction is sent as the response DeprovisionInstance requests.
+func (r *UpdateInstanceReaction) react(_ *v2.UpdateInstanceRequest) (*v2.UpdateInstanceResponse, error) {
+	if r == nil {
+		return nil, UnexpectedActionError()
+	}
+	return r.Response, r.Error
+}
+
+type DynamicUpdateInstanceReaction func(*v2.UpdateInstanceRequest) (*v2.UpdateInstanceResponse, error)
+
+func (r DynamicUpdateInstanceReaction) react(req *v2.UpdateInstanceRequest) (*v2.UpdateInstanceResponse, error) {
+	return r(req)
+}
+
+// DeprovisionReactionInterface is sent as the response DeprovisionInstance requests.
+type DeprovisionReactionInterface interface {
+	react(*v2.DeprovisionRequest) (*v2.DeprovisionResponse, error)
+}
+
 type DeprovisionReaction struct {
-	Response *v2.DeprovisionResponse
-	Error    error
+	Response     *v2.DeprovisionResponse
+	Error        error
+	FuncReaction func(*v2.DeprovisionRequest) (*v2.DeprovisionResponse, error)
```

Comment:
This (and the rest of the `FuncReactions` for the other types) don't seem to be used anywhere. What are they intended for, or can they be removed?

Reply:
They were erroneously left around from an earlier design.

Reply:
@staebler Looks like you missed the ones for `DeprovisionReaction` and `PollLastOperationReaction`